### ISSUE TITLE
Oracle not handling extra bind variable; error in tables 

### ISSUE THREAD

**Issue by Freewilly3d** [chrthomsen/pygrametl#9]
I dont' know if I'm missing something or not.  I'm experiencing some issues because Oracle (cx_Oracle) queries fail when there are additional bind variables that are not used in the query with the following error:  cx_Oracle.DatabaseError: ORA-01036: illegal variable name/number?

Is there a way around this?
I was able to fix the straight table.lookup pretty easy by creating a temp string with just the lookupatts.

Thanks

**Comment by chrthomsen** [comment]
Hello,
Thanks for the report. I don't have an Oracle installation where I can try this out. But just to make sure I understand it correctly: If the lookupatts are ['a', 'b'], you get the error if the row (i.e., dict) to lookup contains any more keys than 'a' and 'b'? Do you create a temp *string* to get around this?

**Comment by Freewilly3d** [comment]
Yes, I’ve started inserting something like this into table.py for table types I’m currently using to deal with it.  

 

Line 226

templist = self.lookupatts
temp = {key: row[key] for key in templist}

 

384

templist = self.attributes + self.lookupatts
templist.append(self.key)
temp = {key: row[key] for key in templist}

 

 

1721

templist = self.measures + self.keyrefs
temp = {key: row[key] for key in templist}

 

 

I’ve been making some other changes and was going to make a pull request after I’ve tested them:

 

1)  I’ve started added error handling (try/except) to all the query executions to help speed debugging. When the SQL is for inserts I added a function to convert the sql and the dictionary into actual sql.  This allowed me to quickly test against the database using a query tool to quickly see what’s going on.  Oracle error messages are cryptic and this helped speed things along.

2)  Temporarily I added to code @ line 101 to data sources to deal with inserting blanks into number columns.  Oracle was returning an invalid number error on csv columns with no value when the columns where defined as a number type.  I haven’t analyzed what was happening here and this may need to change to null instead of 0.  There are a couple different camps around best practices for handling numeric columns that have no value vs columns that have bad values.

Standke 2016-11-10 for csv handle missing data values for numbers
            try:
                row[att] = func(row[att])
            except:
                row[att] = 0
        return row

 

3)  I some had some issues with dates and found that I needed to convert them to python dates with a little function to insert those rows:
def convertCHARinDate(inDate):
    tempdate = datetime.datetime.strptime(row[inDate],"%m/%d/%y").date()
    return tempdate
 

row['calendar_date'] = convertCHARinDate('weekenddt')
row['weekenddt_sid'] = timeday_dimension.lookup(row)

 
4)  I’d also like to change the file name on typedcsvsource to f instead of csvfile so it matches the CSVSource or vice versa so there consistent.  That was a little confusing to me at first when I was working with the beginner sample.
 

 

Otherwise, this is an awesome framework.  I’m working on a POC for a project and was able to load the actual dimensional mart with test data in a day.  The POS includes 4 fact tables 7 dimensions.

 

 

Bill

Bill Standke

Business Intelligence, Enterprise Architecture, Project Management

Mobile: 612-325-9222  | bill@standke.us

 <http://www.linkedin.com/in/billstandke/>

**Comment by Freewilly3d** [comment]
Yes, I’ve started inserting something like this into table.py for table types I’m currently using to deal with it.  

 

Line 226

templist = self.lookupatts
temp = {key: row[key] for key in templist}

 

384

templist = self.attributes + self.lookupatts
templist.append(self.key)
temp = {key: row[key] for key in templist}

 

 

1721

templist = self.measures + self.keyrefs
temp = {key: row[key] for key in templist}

 

 

I’ve been making some other changes and was going to make a pull request after I’ve tested them:

 

1)  I’ve started added error handling (try/except) to all the query executions to help speed debugging. When the SQL is for inserts I added a function to convert the sql and the dictionary into actual sql.  This allowed me to quickly test against the database using a query tool to quickly see what’s going on.  Oracle error messages are cryptic and this helped speed things along.

2)  Temporarily I added to code @ line 101 to data sources to deal with inserting blanks into number columns.  Oracle was returning an invalid number error on csv columns with no value when the columns where defined as a number type.  I haven’t analyzed what was happening here and this may need to change to null instead of 0.  There are a couple different camps around best practices for handling numeric columns that have no value vs columns that have bad values.

Standke 2016-11-10 for csv handle missing data values for numbers
            try:
                row[att] = func(row[att])
            except:
                row[att] = 0
        return row

 

3)  I some had some issues with dates and found that I needed to convert them to python dates with a little function to insert those rows:
def convertCHARinDate(inDate):
    tempdate = datetime.datetime.strptime(row[inDate],"%m/%d/%y").date()
    return tempdate
 

row['calendar_date'] = convertCHARinDate('weekenddt')
row['weekenddt_sid'] = timeday_dimension.lookup(row)

 
4)  I’d also like to change the file name on typedcsvsource to f instead of csvfile so it matches the CSVSource or vice versa so there consistent.  That was a little confusing to me at first when I was working with the beginner sample.
 

 

Otherwise, this is an awesome framework.  I’m working on a POC for a project and was able to load the actual dimensional mart with test data in a day.  The POS includes 4 fact tables 7 dimensions.

 

 

Bill

 

Bill Standke

Business Intelligence, Enterprise Architecture, Project Management

Mobile: 612-325-9222  | bill@standke.us

 <http://www.linkedin.com/in/billstandke/>

**Comment by chrthomsen** [comment]
Thanks for the details. Our initial thought about the problem with cx_oracle is that it perhaps could be better to deal with this in ConnectionWrapper (such that we avoid creating new dicts for all other drivers). We already have some functions to translate into the different formats used by PEP249 drivers and we could extend that to also make new dicts when needed.

About 2), you could make a function to convert values and then use that in TypedCSVSource's casts argument:
myint = lambda x: pygrametl.getint(x, 0)
src = TypedCSVSource(..., casts={'col':myint}, ...)

The we avoid to hardcode 0 or None into the pygrametl code.

About 4), we agree that they should be similar. The Python doc says "csvfile" should be used, but the implementations (at least those we have tried) use "f". We will use "f".

Best regards,
Søren and Christian
